File: victron-monitor/web/lib/server/analytics.ts
import 'server-only';
import { PostHog } from 'posthog-node';

// Server-side half of analytics (2026-09-19) — the client-side snippet
// (components/analytics/PostHogProvider.tsx) only ever sees a browser
// event; "how many subscribe, how many unsubscribe" are backend actions
// with no page load of their own, so they're captured from here instead,
// at the exact point each one actually succeeds — see
// `lib/server/signup.ts` (signup_request_submitted) and
// `app/api/billing/cancel/route.ts` (subscription_cancelled).
//
// `subscription_started` is NOT captured from this app at all — that
// resolution happens inside vrm_api's billing reconciliation, which has
// no single clean call site in Next.js to hook (a webhook retry, the
// daily reconcile-due sweep, and a customer's own subscribe/cancel/
// change action all funnel through the same `reconcile_customer()`).
// It's captured from `vrm_api/billing.py:apply_entitlements()` instead
// (Python's own `posthog` SDK, `vrm_api/analytics.py`), at the one
// `provisioning_state` transition (`pending_subscription` -> `active`)
// that's structurally guaranteed to happen exactly once per customer —
// see that call site's own comment for the full reasoning.
//
// Module-level singleton, not one client per request — PostHog's own
// Node docs recommend this for exactly the reason `flushAt`/
// `flushInterval` below exist: a serverless function can be frozen or
// recycled between requests, so this also calls `.shutdown()` isn't
// wired up per-request (would defeat the singleton); `flushAt: 1` sends
// each event immediately instead of batching, trading a little latency
// for not losing events if the process is torn down before a batch
// flushes.
let client: PostHog | null | undefined;

export function getServerPostHog(): PostHog | null {
  if (client !== undefined) return client;
  const key = process.env.NEXT_PUBLIC_POSTHOG_KEY;
  client = key
    ? new PostHog(key, {
        host: process.env.NEXT_PUBLIC_POSTHOG_HOST || 'https://us.i.posthog.com',
        flushAt: 1,
        flushInterval: 0,
      })
    : null;
  return client;
}

/**
 * Fire-and-forget capture — every call site here is on a success path
 * that must never fail or delay because analytics is unreachable
 * (same "this is a side effect, not a dependency" rule the rest of this
 * app already applies to its own non-critical integrations). `distinctId`
 * is the customer's email, lowercased — the same identity `vrm.customers.
 * auth_email` uses, so a later `posthog.identify()` call (once a real
 * customer/session exists) merges into the same person instead of
 * creating a second, orphaned one.
 */
export function captureServerEvent(distinctId: string, event: string, properties?: Record<string, unknown>): void {
  const ph = getServerPostHog();
  if (!ph) return;
  try {
    ph.capture({ distinctId, event, properties });
  } catch (err) {
    console.error(`captureServerEvent: failed to capture "${event}"`, err);
  }
}

File: vrm_api/analytics.py
from __future__ import annotations
"""
vrm_api's own half of the analytics setup (2026-09-19) — mirrors
`victron-monitor/web/lib/server/analytics.ts` on the Next.js side, same
project, same PostHog key, just the Python SDK instead of posthog-node.

Exists for exactly one event this app fires: `subscription_started`, from
`billing.py:apply_entitlements()`, at the one place `provisioning_state`
is promoted from `'pending_subscription'` to `'active'` — see that call
site's own comment for why THAT specific transition is safe to treat as
"fires exactly once, ever, per customer" (PLAN_PHASE16.md §4.5 rule 8:
promotion is one-way and never reverts), unlike almost anything else in
`reconcile_customer()`, which re-reads and rewrites the mirror wholesale
on every call — a webhook retry, the daily reconcile-due sweep, and a
customer's own subscribe/cancel/change action all funnel through it, so
naively firing an event anywhere else in that function would double- or
triple-count the same real-world event.

The `trial_started`/`signup_request_submitted`/`subscription_cancelled`
events all live on the Next.js side instead — this file is not a general
analytics module for vrm_api, just this one funnel step.
"""
import logging
import os

logger = logging.getLogger("vrm_api.analytics")

_client = None
_client_initialized = False


def _get_client():
    """Lazy, module-level singleton — mirrors analytics.ts's own reasoning:
    constructed once per process, not once per call, but never at import
    time either (importing this module must never require the env var to
    already be set, e.g. in a test run)."""
    global _client, _client_initialized
    if _client_initialized:
        return _client
    _client_initialized = True

    key = os.environ.get("NEXT_PUBLIC_POSTHOG_KEY")
    if not key:
        _client = None
        return None

    try:
        from posthog import Posthog
        _client = Posthog(
            key,
            host=os.environ.get("NEXT_PUBLIC_POSTHOG_HOST") or "https://us.i.posthog.com",
        )
    except Exception:  # noqa: BLE001 — analytics must never break billing
        logger.exception("analytics: failed to construct PostHog client")
        _client = None
    return _client


def capture_server_event(distinct_id: str, event: str, properties: dict | None = None) -> None:
    """Fire-and-forget — every call site is on a billing success path that
    must never fail, slow down, or roll back over an unreachable analytics
    endpoint. `distinct_id` should be the customer's email where available
    (the same identity `vrm.customers.auth_email` / the Next.js side's own
    events use), falling back to `customer_id` only when no email is on
    hand at the call site."""
    client = _get_client()
    if client is None:
        return
    try:
        client.capture(distinct_id=distinct_id, event=event, properties=properties or {})
    except Exception:  # noqa: BLE001 — see module/function docstrings
        logger.exception("analytics: failed to capture %r", event)
